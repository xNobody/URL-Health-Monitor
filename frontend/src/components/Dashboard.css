body {
    margin: 0;
    padding: 0;
    background-color: #130f40;
    background-position: center;
    font-family: sans-serif;
    color: #2c3e50;
}

.dashboard-container {
    width: 90%;
    max-width: 1200px;
    background-color: #ecf0f1;
    margin: 50px auto;
    padding: 20px;
    border: 5px solid #34495e;
    border-radius: 10px;
    box-shadow: 0px 0px 5px #000;
}

.dashboard-container h1 {
    text-align: center;
    font-size: 28px;
    margin-bottom: 20px;
}

.dashboard-container .monitor {
    margin-bottom: 20px;
    padding: 10px;
    border-bottom: 1px solid #34495e;
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: relative;
}

.dashboard-container .monitor h2 {
    font-size: 22px;
    margin: 0;
}

.dashboard-container .monitor p {
    margin: 5px 0;
}

.dashboard-container .monitor .monitor-details {
    flex-grow: 1;
}

.dashboard-container .monitor .monitor-buttons {
    display: flex;
    align-items: center;
    position: absolute;
    right: 10px;
    top: 10px;
}

.dashboard-container .monitor button {
    border: none;
    outline: none;
    height: 40px;
    background: linear-gradient(45deg, #2c3e50, #130f40);
    color: #ffffff;
    font-size: 16px;
    font-weight: bold;
    border-radius: 20px;
    box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
    transition: background 0.5s, color 0.5s, box-shadow 0.5s;
    margin: 0 5px;
}

.dashboard-container .monitor button:hover {
    cursor: pointer;
    background: linear-gradient(45deg, #130f40, #2c3e50);
    color: #ffffff;
    box-shadow: 0px 0px 15px rgba(0, 0, 0, 0.7);
}

.edit-button {
    background-color: #28a745;
}

.edit-button:hover {
    background-color: #218838;
}

.delete-button {
    background-color: #dc3545;
}

.delete-button:hover {
    background-color: #c82333;
}

.watermark-message {
    text-align: center;
    color: #888;
    font-size: 18px;
    margin-top: 50px;
}