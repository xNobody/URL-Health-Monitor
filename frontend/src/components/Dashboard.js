import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { Line } from 'react-chartjs-2';
import { Chart as ChartJS, ArcElement, Tooltip, Legend, LineElement, CategoryScale, LinearScale, PointElement } from 'chart.js';
import './Dashboard.css';

ChartJS.register(ArcElement, Tooltip, Legend, LineElement, CategoryScale, LinearScale, PointElement);

const Dashboard = () => {
  const [monitors, setMonitors] = useState([]);
  const [historyData, setHistoryData] = useState({});
  const [visibleCharts, setVisibleCharts] = useState({});

  useEffect(() => {
    const token = localStorage.getItem('token');
    if (!token) {
      console.error('No token found');
      return;
    }
    console.log('Token:', token); // Debugging line
    axios.get('https://localhost:3000/api/url_monitors', {
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(response => {
        console.log('Monitors response:', response.data); // Debugging line
        const monitors = response.data;
        setMonitors(monitors);

        monitors.forEach(monitor => {
          axios.get(`https://localhost:3000/api/url_monitors/${monitor.id}/history`, {
            headers: {
              Authorization: `Bearer ${token}`
            }
          })
            .then(historyResponse => {
              setHistoryData(prevHistoryData => ({
                ...prevHistoryData,
                [monitor.id]: historyResponse.data
              }));
            })
            .catch(error => console.error(`Error fetching history for monitor ${monitor.id}:`, error));
        });
      })
      .catch(error => console.error('Error fetching monitors:', error));
  }, []);

  const getChartData = (monitorId) => {
    const history = historyData[monitorId] || [];
    return {
      labels: history.map(entry => new Date(entry.checked_at).toLocaleString()),
      datasets: [
        {
          label: 'Response Time (ms)',
          data: history.map(entry => entry.response_time),
          fill: false,
          backgroundColor: 'rgba(75,192,192,0.4)',
          borderColor: 'rgba(75,192,192,1)',
        },
      ],
    };
  };

  const toggleChartVisibility = (monitorId) => {
    setVisibleCharts(prevVisibleCharts => ({
      ...prevVisibleCharts,
      [monitorId]: !prevVisibleCharts[monitorId]
    }));
  };

  return (
    <div className="dashboard-container">
      <h1>Monitored URLs</h1>
      {monitors.map(monitor => (
        <div key={monitor.id} className="monitor">
          <h2>{monitor.name}</h2>
          <p>URL: {monitor.url}</p>
          <p>Status: {monitor.status || 'Unknown'}</p>
          <p>Last Checked: {monitor.last_checked_at ? new Date(monitor.last_checked_at).toLocaleString() : 'Never'}</p>
          <button onClick={() => toggleChartVisibility(monitor.id)}>
            {visibleCharts[monitor.id] ? 'Hide Chart' : 'Show Chart'}
          </button>
          {visibleCharts[monitor.id] && <Line data={getChartData(monitor.id)} />}
        </div>
      ))}
    </div>
  );
};

export default Dashboard;
