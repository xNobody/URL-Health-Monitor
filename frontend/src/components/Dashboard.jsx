import React, { useEffect, useState, useCallback } from 'react';
import axios from 'axios';
import { Line } from 'react-chartjs-2';
import { Chart as ChartJS, ArcElement, Tooltip, Legend, LineElement, CategoryScale, LinearScale, PointElement } from 'chart.js';
import MonitorForm from './MonitorForm';
import './Dashboard.css';

ChartJS.register(ArcElement, Tooltip, Legend, LineElement, CategoryScale, LinearScale, PointElement);

const Dashboard = () => {
  const [monitors, setMonitors] = useState([]);
  const [filteredMonitors, setFilteredMonitors] = useState([]);
  const [historyData, setHistoryData] = useState({});
  const [visibleCharts, setVisibleCharts] = useState({});
  const [editingMonitor, setEditingMonitor] = useState(null);
  const [addingMonitor, setAddingMonitor] = useState(false);
  const [filter, setFilter] = useState('');
  const [sortOrder, setSortOrder] = useState('asc');

  useEffect(() => {
    const token = localStorage.getItem('token');
    if (!token) {
      console.error('No token found');
      return;
    }
    axios.get('https://localhost:3000/api/url_monitors', {
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(response => {
        const monitors = response.data;
        setMonitors(monitors);
        setFilteredMonitors(monitors);

        monitors.forEach(monitor => {
          axios.get(`https://localhost:3000/api/url_monitors/${monitor.id}/history`, {
            headers: {
              Authorization: `Bearer ${token}`
            }
          })
            .then(historyResponse => {
              setHistoryData(prevHistoryData => ({
                ...prevHistoryData,
                [monitor.id]: historyResponse.data
              }));
            })
            .catch(error => console.error(`Error fetching history for monitor ${monitor.id}:`, error));
        });
      })
      .catch(error => console.error('Error fetching monitors:', error));
  }, []);

  const handleFilterAndSort = useCallback(() => {
    let filtered = monitors.filter(monitor =>
      monitor.name.toLowerCase().includes(filter.toLowerCase()) ||
      monitor.url.toLowerCase().includes(filter.toLowerCase())
    );

    filtered.sort((a, b) => {
      if (sortOrder === 'asc') {
        return a.name.localeCompare(b.name);
      } else {
        return b.name.localeCompare(a.name);
      }
    });

    setFilteredMonitors(filtered);
  }, [monitors, filter, sortOrder]);

  useEffect(() => {
    handleFilterAndSort();
  }, [filter, sortOrder, monitors, handleFilterAndSort]);

  const getChartData = (monitorId) => {
    const history = historyData[monitorId] || [];
    return {
      labels: history.map(entry => new Date(entry.checked_at).toLocaleString()),
      datasets: [
        {
          label: 'Response Time (ms)',
          data: history.map(entry => entry.response_time),
          fill: false,
          backgroundColor: 'rgba(75,192,192,0.4)',
          borderColor: 'rgba(75,192,192,1)',
        },
      ],
    };
  };

  const toggleChartVisibility = (monitorId) => {
    setVisibleCharts(prevVisibleCharts => ({
      ...prevVisibleCharts,
      [monitorId]: !prevVisibleCharts[monitorId]
    }));
  };

  const handleDelete = (monitorId) => {
    const token = localStorage.getItem('token');
    axios.delete(`https://localhost:3000/api/url_monitors/${monitorId}`, {
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(() => {
        setMonitors(monitors.filter(monitor => monitor.id !== monitorId));
      })
      .catch(error => console.error(`Error deleting monitor ${monitorId}:`, error));
  };

  const handleEdit = (monitorId) => {
    const monitor = monitors.find(m => m.id === monitorId);
    setEditingMonitor(monitor);
  };

  const handleSave = (updatedMonitor) => {
    const token = localStorage.getItem('token');
    axios.put(`https://localhost:3000/api/url_monitors/${updatedMonitor.id}`, { url_monitor: updatedMonitor }, {
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(response => {
        setMonitors(monitors.map(m => (m.id === updatedMonitor.id ? response.data : m)));
        setEditingMonitor(null);
      })
      .catch(error => {
        console.error('Error updating monitor:', error);
      });
  };

  const handleAdd = (newMonitor) => {
    const token = localStorage.getItem('token');
    axios.post('https://localhost:3000/api/url_monitors', { url_monitor: newMonitor }, {
      headers: {
        Authorization: `Bearer ${token}`
      }
    })
      .then(response => {
        setMonitors([...monitors, response.data]);
        setAddingMonitor(false);
      })
      .catch(error => {
        console.error('Error adding monitor:', error);
      });
  };

  const getStatusClass = (status) => {
    if (status === 'up') return 'status-pill status-up';
    if (status === 'down') return 'status-pill status-down';
    return 'status-pill status-unknown';
  };

  return (
    <div className="dashboard-container">
      <h1>Monitored URLs</h1>
      <div className="filter-sort-container">
        <input
          type="text"
          placeholder="Search by Name or URL"
          value={filter}
          onChange={(e) => setFilter(e.target.value)}
        />
        <select value={sortOrder} onChange={(e) => setSortOrder(e.target.value)}>
          <option value="asc">Sort by Name (A-Z)</option>
          <option value="desc">Sort by Name (Z-A)</option>
        </select>
      </div>
      <button onClick={() => setAddingMonitor(true)} className="add-button">Add Monitor</button>
      {filteredMonitors.length === 0 ? (
        <div className="watermark-message">
          <p>No URLs are being monitored. Please add a monitor to get started.</p>
        </div>
      ) : (
        filteredMonitors.map(monitor => (
          <div key={monitor.id} className="monitor">
            <div className="monitor-details">
              <h2>{monitor.name}</h2>
              <p>URL: {monitor.url}</p>
              <p>Status: <span className={getStatusClass(monitor.status)}>{monitor.status || 'Unknown'}</span></p>
              <p>Last Checked: {monitor.last_checked_at ? new Date(monitor.last_checked_at).toLocaleString() : 'Never'}</p>
              <button onClick={() => toggleChartVisibility(monitor.id)}>
                {visibleCharts[monitor.id] ? 'Hide Chart' : 'Show Chart'}
              </button>
              {visibleCharts[monitor.id] && <Line data={getChartData(monitor.id)} />}
            </div>
            <div className="monitor-buttons">
              <button onClick={() => handleEdit(monitor.id)} className="edit-button">Edit</button>
              <button onClick={() => handleDelete(monitor.id)} className="delete-button">Remove</button>
            </div>
          </div>
        ))
      )}
      {editingMonitor && (
        <MonitorForm
          initialData={editingMonitor}
          onSubmit={handleSave}
          onClose={() => setEditingMonitor(null)}
        />
      )}
      {addingMonitor && (
        <MonitorForm
          onSubmit={handleAdd}
          onClose={() => setAddingMonitor(false)}
        />
      )}
    </div>
  );
};

export default Dashboard;
