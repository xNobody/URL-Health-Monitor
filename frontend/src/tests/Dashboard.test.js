import React from 'react';
import { render } from '@testing-library/react';
import Dashboard from '../components/Dashboard';

describe('Dashboard Component Snapshot', () => {
  test('matches the snapshot', () => {
    const { asFragment } = render(<Dashboard />);
    expect(asFragment()).toMatchSnapshot();
  });
});