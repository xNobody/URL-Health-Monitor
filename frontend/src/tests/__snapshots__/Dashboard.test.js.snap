// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component Snapshot matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="dashboard-container"
  >
    <h1>
      Monitored URLs
    </h1>
    <div
      class="filter-sort-container"
    >
      <input
        placeholder="Search by Name or URL"
        type="text"
        value=""
      />
      <select>
        <option
          value="asc"
        >
          Sort by Name (A-Z)
        </option>
        <option
          value="desc"
        >
          Sort by Name (Z-A)
        </option>
      </select>
    </div>
    <div
      class="watermark-message"
    >
      <p>
        No URLs are being monitored. Please add a monitor to get started.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
